spacy_ray/optimizer.py

```diff
@@ -0,0 +1,75 @@
+"""Allreduce distributed training with Ray."""
+
+from wasabi import msg
+from spacy import util
+from typing import Tuple
+from thinc.types import FloatsXd
+
+cp = None
+nccl = None
+
+try:
+    import cupy as cp
+    from cupy.cuda import nccl
+except ImportError:
+    msg.fail("Need to `pip install cupy-[driver version]` to use "
+             "multi-gpu distributed training.")
+
+
+def create_optimizer(config_path):
+    msg.info(f"Loading config from: {config_path}")
+    config = util.load_config(config_path, create_objects=False)
+    util.fix_random_seed(config["training"]["seed"])
+    config = util.load_config(config_path, create_objects=True)
+    training = config["training"]
+    return training["optimizer"]
+
+
+class RayNCCLWorker:
+    def __init__(self, rank, world_size):
+        self.rank = rank
+        self.world_size = world_size
+        self.unique_id = nccl.get_unique_id()
+
+    def initialize(self, head_id):
+        self.communicator = nccl.NcclCommunicator(self.world_size, head_id,
+                                                  self.rank)
+
+    def get_unique_id(self):
+        return self.unique_id
+
+    def execute(self, fn):
+        return fn(self)
+
+
+class AllreduceOptimizer:
+    def __init__(self, config_path, communicator):
+        self.optimizer = create_optimizer(config_path)
+        self.communicator = communicator
+        self.weights_synced = set()
+
+    def allreduce(self, tensor):
+        self.communicator.allReduce(tensor.data.ptr, tensor.data.ptr,
+                                    tensor.size, nccl.NCCL_FLOAT32,
```

Comment:
infer the date type to use from the input tensor instead of using nccl.NCCL_FLOAT32